tests/unittests/test_runs/test_simple.py

```diff
@@ -0,0 +1,282 @@
+    'openstack/2012-08-10/meta_data.json': json.dumps(OSTACK_META),
+    'openstack/2012-08-10/user_data': USER_DATA,
+    'openstack/content/0000': CONTENT_0,
+    'openstack/content/0001': CONTENT_1,
+    'openstack/latest/meta_data.json': json.dumps(OSTACK_META),
+    'openstack/latest/user_data': USER_DATA,
+    'openstack/latest/network_data.json': json.dumps(NETWORK_DATA),
+    'openstack/2015-10-15/meta_data.json': json.dumps(OSTACK_META),
+    'openstack/2015-10-15/user_data': USER_DATA,
+    'openstack/2015-10-15/network_data.json': json.dumps(NETWORK_DATA)}
+
+
+class TestSimpleRun(helpers.FilesystemMockingTestCase):
+    def setUp(self):
+        super(TestSimpleRun, self).setUp()
+        self.new_root = tempfile.mkdtemp()
+        self.tmp = self.tmp_dir()
+        self.sys_cfg = {'datasource': {'ConfigDrive': {'dsmode': 'local'}}}
+        self.paths = cloud_helpers.Paths(
+            {'cloud_dir': self.tmp, 'run_dir': self.tmp})
+        self.ds = ds.DataSourceConfigDrive
+
+    def _patchIn(self, root):
+        self.patchOS(root)
+        self.patchUtils(root)
+
+    @helpers.mock.patch('cloudinit.sources.find_source')
+    def test_invalid_config_drive_subsequent_boot(self, mock_ds):
+        self.addCleanup(shutil.rmtree, self.new_root)
+        self.replicateTestRoot('simple_ubuntu', self.new_root)
+        cfg = {
+            'datasource_list': ['ConfigDrive'],
+            'cloud_init_modules': ['write-files'],
+            'system_info': {'paths': {'run_dir': self.new_root}}
+        }
+        orig_find_devs_with = util.find_devs_with
+        try:
+            # dont' try to lookup for CDs
+            util.find_devs_with = lambda path: []
+            dsrc = self.ds(sys_cfg=self.sys_cfg, distro=None, paths=self.paths)
+            populate_dir(dsrc.seed_dir, CFG_DRIVE_FILES_V2)
+            found = dsrc._get_data()
+            self.assertTrue(found)
+        finally:
+            util.find_devs_with = orig_find_devs_with
+
+        helpers.readResource('user_data.1.txt')
+        cloud_cfg = safeyaml.dumps(cfg)
+        util.ensure_dir(os.path.join(self.new_root, 'etc', 'cloud'))
+        util.write_file(os.path.join(self.new_root, 'etc',
+                                     'cloud', 'cloud.cfg'), cloud_cfg)
+        self._patchIn(self.new_root)
+        # simulate fist boot with config drive
+        mock_ds.return_value = (dsrc, 'ConfigDrive')
+        initer = stages.Init()
+        initer.read_cfg()
+        initer.initialize()
+        initer.fetch()
+        initer.instancify()
+        initer.update()
+        # simulate config drive removal and reboot
+        util.del_file(".instance-id")
+        mock_ds.reset_mock()
+        msg = ("Did not find any data source")
+        mock_ds.side_effect = DataSourceNotFoundException(msg)
+        initer = stages.Init()
+        initer.read_cfg()
+        initer.initialize()
+        initer.fetch()
+        initer.instancify()
+        initer.update()
```

Comment:
Shouldn't we check that the instance-id on a ds object from the "second" boot has the same instance-id value from the first boot?